Theatre of Blood Session stats plugin

TheStonedTurtle opened the pull request:
Adds a plugin that tracks session stats for Theatre of Blood. This only tracks for the local player.

**NOTE: This data is lost when the client is closed**

### Tracked Stats

<details>
<p>

For each run, or **Attempt**, at the Theatre of Blood the following information will be tracked:

* Damage Dealt *(approximation)*
* Damage Taken
* Death Count
* Successfully Completed

Information is tracked by room, displaying the below information upon completion.
* Damage Dealt *(approximation)*
* Damage Taken

An Act is considered finished when you enter the next room, there is a total of 6 acts.

</p>
</details>

### Media

<details>
<p>

### In-Game messages:
![](https://i.imgur.com/Hhv1OKV.png)

</p>
</details>

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/overlays/AttemptOverlay.java:
```diff
@@ -0,0 +1,98 @@
+import net.runelite.client.ui.overlay.components.PanelComponent;
+import net.runelite.client.ui.overlay.components.TitleComponent;
+
+public class AttemptOverlay extends Overlay
```
These shouldn't be separate overlays

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/data/Attempt.java:
```diff
@@ -0,0 +1,71 @@
+import lombok.Setter;
+
+@Getter
+public class Attempt
```
All of these data structures feel extraneous.

TheStonedTurtle replied:
I'm trying to track by room and by attempt and this seemed like the easiest way, what would you recommend I do instead?

raiyni commented:
Each attempt looks to be a sum of each room. Each room seems to historically current room. It also doesn't look like you are doing any advanced per room statistics. So I would guess a couplve v with enum keys would replace a lot of this boiler plate.

MagicfTail commented:
Issues found from testing so far:

1. The raids completion % shows 1 instead of 100, idk if it's the same without 100% completion rate.
2. The death counter adds deaths to session doing 0 death raids, no deaths are added during the rooms themselves, so idk where they are coming from. 2 were added the first raid, and thereafter 1 every raid.

MagicfTail commented:
3. The raid overlays show outside the rooms/theatre of blood area

TheStonedTurtle commented:
> 
>     2. The death counter adds deaths to session doing 0 death raids, no deaths are added during the rooms themselves, so idk where they are coming from. 2 were added the first raid, and thereafter 1 every raid.

Deaths are currently being tracked by whenever the ToB Varbit changes to 3, it must change to 3 for some other reason as well, I'll have to change the way this is tracked.

TheStonedTurtle commented:
@raiyni I've removed the overlays entirely and tried to simplify. Let me know if you see anything else.

@MagicfTail I believe I've fixed those issues, if you don't mind testing again lmk if you have any issues.

Thanks guys

MagicfTail commented:
Completion % get's rounded to 0

TheStonedTurtle commented:
> 
> 
> Completion % get's rounded to 0

I'm not sure how this is happening.. `(completions / attempts.size()) * 100` is what I'm doing to determine the percentage and I'm formatting it via `new DecimalFormat("#,###");'. Any ideas?

TheStonedTurtle commented:
Removed the damage capping functionality as it messed with AOE attacks and wasn't that accurate anyway.

NathenSample commented on runelite-client/src/main/java/net/runelite/client/plugins/theatreofblood/TheatreOfBloodPlugin.java:
```diff
@@ -0,0 +1,472 @@
+	private void stateChanged(int old)
+	{
+		// TODO: Figure out a way to determine if they are logging back into a raid
+		// 0=default | 1=party | 2=inside/spectator | 3=dead-spectator
```
Just to re-iterate the above - Once you're at the stage you're describing what your ints mean with comments, use an enum ;)

TheStonedTurtle commented:
Working on making the plugin more generic for expand-ability, see #6452
